src/OrleansRuntime/MultiClusterNetwork/MultiClusterOracle.cs

```diff
@@ -0,0 +1,766 @@
+        private TimeSpan RandomizeTimespan(TimeSpan value)
+        {
+            return TimeSpan.FromMilliseconds(value.TotalMilliseconds * (0.9 + (random.NextDouble() * 0.1)));
+        }
+
+        public bool IsFunctionalClusterGateway(SiloAddress siloAddress)
+        {
+            GatewayEntry g;
+            return localData.Current.Gateways.TryGetValue(siloAddress, out g) 
+                && g.Status == GatewayStatus.Active;
+        }
+
+        public IEnumerable<string> GetActiveClusters()
+        {
+            var clusters = localData.Current.Gateways.Values
+                 .Where(g => g.Status == GatewayStatus.Active)
+                 .Select(g => g.ClusterId)
+                 .Distinct();
+
+            return clusters;
+        }
+
+        public IEnumerable<GatewayEntry> GetGateways()
+        {
+            return localData.Current.Gateways.Values;
+        }
+
+        public SiloAddress GetRandomClusterGateway(string cluster)
+        {
+            var activeGateways = this.localData.Current.Gateways.Values
+                .Where(gw => gw.ClusterId == cluster && gw.Status == GatewayStatus.Active)
+                .Select(gw => gw.SiloAddress)
+                .ToList();
+
+            if (activeGateways.Count == 0)
+                return null;
+
+            return activeGateways[random.Next(activeGateways.Count)];
+        }
+
+        public MultiClusterConfiguration GetMultiClusterConfiguration()
+        {
+            return localData.Current.Configuration;
+        }
+
+        public async Task InjectMultiClusterConfiguration(MultiClusterConfiguration config)
+        {
+            this.injectedConfig = config;
+            PushChanges();
+
+            // wait for the gossip channel tasks and aggregate exceptions
+            var currentChannelTasks = this.channelTasks.Values.ToList();
+            await Task.WhenAll(currentChannelTasks.Select(ct => ct.Task));
+
+            var exceptions = currentChannelTasks
+                .Where(ct => ct.LastException != null)
+                .Select(ct => ct.LastException)
+                .ToList();
+
+            if (exceptions.Count > 0)
+                throw new AggregateException(exceptions);
+        }
+
+        public void SiloStatusChangeNotification(SiloAddress updatedSilo, SiloStatus status)
+        {
+            // any status change can cause changes in gateway list
+            PushChanges();
+        }
+
+        public async Task Start(ISiloStatusOracle oracle)
+        {
+            logger.Info(ErrorCode.MultiClusterNetwork_Starting, "MultiClusterOracle starting on {0}, Severity={1} ", Silo, logger.SeverityLevel);
+            try
+            {
+                if (string.IsNullOrEmpty(clusterId))
+                    throw new OrleansException("Internal Error: missing cluster id");
+
+                this.siloStatusOracle = oracle;
+
+                this.siloStatusOracle.SubscribeToSiloStatusEvents(this);
```

Comment:
It is probably safe, but I agree that it would make more sense to subscribe a bit later.
I will move this down, to right before PushChanges().